pyccl/pk2d.py

```diff
@@ -248,13 +253,145 @@ def eval_dlogpk_dlogk(self, k, a, cosmo):
 
         return f
 
+    def get_spline_arrays(self):
+        """Get the spline data arrays.
+
+        Returns:
+            a_arr: array_like
+                Array of scale factors.
+            lk_arr: array_like
+                Array of logarithm of wavenumber k.
+            pk_arr: array_like
+                Array of the power spectrum P(k, z). The shape
+                is (a_arr.size, lk_arr.size).
+        """
+        if not self.has_psp:
+            raise ValueError("Pk2D object does not have data.")
+
+        a_arr, lk_arr, pk_arr = _get_spline2d_arrays(self.psp.fka)
+        if self.psp.is_log:
+            pk_arr = np.exp(pk_arr)
+
+        return a_arr, lk_arr, pk_arr
+
     def __del__(self):
         """Free memory associated with this Pk2D structure
         """
         if hasattr(self, 'has_psp'):
             if self.has_psp and hasattr(self, 'psp'):
                 lib.f2d_t_free(self.psp)
 
+    def _check_Pk2D_compatability(self, other):
+        if not isinstance(other, Pk2D):
+            raise ValueError("Binary operator of Pk2D objects is only defined "
+                             "for other Pk2D objects.")
+        if not self.has_psp or not other.has_psp:
+            raise ValueError("Pk2D object does not have data.")
+        if (self.psp.lkmin < other.psp.lkmin
+                or self.psp.lkmax > other.psp.lkmax):
+            raise ValueError("The 2nd operand has its data defined over a "
+                             "smaller k range than the 1st operand.")
+        if (self.psp.amin < other.psp.amin
+                or self.psp.amax > other.psp.amax):
+            raise ValueError("The 2nd operand has its data defined over a "
+                             "smaller a range than the 1st operand.")
```

Comment:
I see the point of throwing an exception whenever one operand has less support than the other operand (instead of, for example, extrapolating). But is there a particular reason this operation is not commutative? Could we not operate on the one with the widest support, in both `a` and `k`?

Reply:
Except if there is a very particular reason where, if we have `B` with narrower support than `A`, we would expect B += A to return `B` with the same support as the original `B`?

Reply:
My issue is that these operations would change the internal state of the Pk2D object in a way that might be opaque to the user. For example, a user creates a Pk2D object over some known a and k range and then adds other Pk2D objects to it. Unless the user carefully track the support of these objects, they might try to access the Pk2D object at k and a values that were inside the support originally but now have fallen outside of it. 
The alternative would be to disallow any operation if the supports don't match and provide some helper function to create new Pk2D object, given some support ranges.

I see the point about more verbose error messages and have now implemented that.

Reply:
>  This is well-defined. A+B should return a Pk2D with the largest support (that of A). No extrapolation necessary.

I disagree. The operation is only well-defined on the intersection. Outside of that you either need to extrapolate `B` or assume `B` is zero, which leads to discontinuities in the power spectrum, neither of which are desirable.

Reply:
Right, I agree that extrapolation is tricky and should be avoided. But at the same time, I would trust that the user has sampled the `Pk2D` well enough so that we can assume it's zero elsewhere in the case of addition, and one in the case of multiplication. This practice is nothing new to CCL; `Pk2D` can accept a user-defined `pk_arr` where we specify that "if you pass your own Pk array, you are responsible of making sure that it is sufficiently well-sampled.
What is the rationale behind only using the intersection? I think we are being too strict, judging from similar situations where we are lenient and allow users to do what they want.

Reply:
If I'm allowed to weigh in, I'd advocate for Keeping It Simple (S). In this sense I see two options:
1. You only allow operations between pks with exactly the same support and throw an exception otherwise.
2. You only allow to add, to a given pk2d A, another pk2d B that requires no extrapolation (i.e. A fits in B), and the result is only defined on the original pk2d's support (i.e. that of A). I think this is what's currently implemented. This breaks commutativity, but it is the simplest implementation. Also, extrapolation would happen only when leaving known ranges (the ones used to define the first pk2d), and never silently (which would happen if we allow for B to fit into A). This behaviour is also easy to document for users.

If we go with number 2 I'd advocate for throwing a warning if someone tries to add two pk2ds with different supports (but compatible in the sense defined above).

I also think commutativity would be great to have, but we can try to get that in afterwards. The way I see it, most people will use this function to operate on power spectra they have created themselves on the same ranges, so in most situations this discussion will be moot.

Reply:
On point 2, which is what we are discussing, how would the users know that `A + B` gives a `Pk2D` with the common range? Wouldn't it feel more natural to expect the resulting pk to have the widest support of the two?

Suppose we have defined a P(k,a) correction factor which is only relevant in `k∈[0.5, 5.0]`. If we want to use this for all `Pk2D` objects down a pipeline we would have to (a) either define a `Pk2D` with an extreme range which contains the tiny section of useful values we want (b) or manually copy and expand the range for each different `Pk2D` we want to operate on.

By the way, I agree with adding a message warning the users of the narrower support if the two supports don't match.